lib/metasploit/framework/mssql/tdssslproxy.rb

```diff
@@ -0,0 +1,122 @@
+# -*- coding: binary -*-
+
+require 'openssl'
+
+#
+# TDSSSLProxy:
+#
+# SQL Server uses the TDS protocol to transmit data between clients and
+# servers. Of course this sits on top of TCP.
+#
+# By default, the TDS payload is not encrypted. However, if "force
+# encryption" is set under the SQL Server protocol properties, it will
+# use SSL/TLS to encrypt the TDS data. Oddly, the entire TCP stream is
+# not encrypted (as is say for HTTPS), but instead a TDS header is
+# put on the front of the TLS packet. As a result, the full TLS/SSL
+# setup is done within a series of TDS payloads.
+#
+# This "proxy" basically creates a fake SSL endpoint (s2) from which it
+# can add/remove the TDS header as required. This is implemented as a
+# socket pair (think, a bidirectional pipe), where the other end is s1:
+#
+# sslsock <-> s1 <-> s2 <-> tdssock <-> target SQL Server.
+#
+# (tdssock is the reference to the "sock" from the scanner module)
+#
+# TO DO:
+#
+# This enables brute force of a SQL Server which requires encryption.
+# However, future updates will permit any read/write using
+# mssql_send_recv() to use crypto if required and transparently to
+# other MSF developers.
+#
+# Cheers, JH
+
+class TDSSSLProxy
+
+  TYPE_TDS7_LOGIN = 16
+  TYPE_PRE_LOGIN_MESSAGE = 18
+  STATUS_END_OF_MESSAGE = 0x01
+
+  def initialize(sock)
+    @tdssock = sock
+    @s1,@s2 = Rex::Socket.tcp_socket_pair()
+  end
+
+  def cleanup()
+    Thread.kill(@t1)
+  end
+
+  def setup_ssl
+    @t1 = Thread.start { ssl_setup_thread() }
```

Comment:
This should use the ``framework.threads.spawn`` API so we can track the thread status:
```
framework.threads.spawn("MSSQL_TDS_SSL_Proxy", false) {
  ssl_setup_thread()
}
```